File: plugins/google-maps-scraper/scraper.py
#!/usr/bin/env python3
"""
Google Maps Scraper - CLI tool for scraping business data from Google Maps
"""

import argparse
import json
import logging
import re
import sys
import time
from typing import List, Dict, Any, Optional

from playwright.sync_api import sync_playwright, Page


def setup_logging(verbose: bool = False):
    level = logging.DEBUG if verbose else logging.INFO
    logging.basicConfig(
        level=level,
        format='%(asctime)s - %(levelname)s - %(message)s',
    )


def extract_text(page: Page, xpath: str) -> str:
    """Extract text from page using XPath selector."""
    try:
        locator = page.locator(xpath)
        if locator.count() > 0:
            return locator.first.inner_text().strip()
    except Exception as e:
        logging.debug(f"Failed to extract text for xpath {xpath}: {e}")
    return ""


def extract_place_details(page: Page) -> Dict[str, Any]:
    """Extract place details from Google Maps page."""
    selectors = {
        "name": '//h1[contains(@class, "DUwDvf")]',
        "address": '//button[@data-item-id="address"]//div[contains(@class, "fontBodyMedium")]',
        "website": '//a[@data-item-id="authority"]//div[contains(@class, "fontBodyMedium")]',
        "phone": '//button[contains(@data-item-id, "phone:tel:")]//div[contains(@class, "fontBodyMedium")]',
        "rating": '//div[contains(@class, "F7nice")]//span[@aria-hidden="true"]',
        "reviews": '//div[contains(@class, "F7nice")]//span[contains(@aria-label, "reviews")]',
        "type": '//button[contains(@class, "DkEaL")]'
    }
    
    details = {}
    for key, xpath in selectors.items():
        details[key] = extract_text(page, xpath)
    
    # Parse reviews count
    if details["reviews"]:
        match = re.search(r'(\d+)', details["reviews"].replace('\xa0', '').replace(',', ''))
        if match:
            details["reviews_count"] = int(match.group(1))
        else:
            details["reviews_count"] = None
    else:
        details["reviews_count"] = None
    
    # Parse rating
    if details["rating"]:
        try:
            details["rating"] = float(details["rating"].replace(' ', '').replace(',', '.'))
        except ValueError:
            details["rating"] = None
    
    return details


def scrape_google_maps(query: str, limit: int = 10, headless: bool = True) -> List[Dict[str, Any]]:
    """Scrape Google Maps for business data."""
    results = []
    logging.info(f"Scraping Google Maps for: {query}")
    
    with sync_playwright() as p:
        browser = p.chromium.launch(headless=headless)
        context = browser.new_context(
            user_agent="Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36"
        )
        page = context.new_page()
        
        try:
            search_url = f"https://www.google.com/maps/search/{query.replace(' ', '+')}"
            page.goto(search_url, timeout=60000)
            
            # Handle cookie consent
            try:
                consent = page.locator('//button[@aria-label="Accept all"] | //button[contains(., "Accept all")]').first
                if consent.is_visible(timeout=5000):
                    consent.click()
                    logging.info("Accepted cookies")
            except:
                pass

            # Check for single result page
            if page.locator('//h1[contains(@class, "DUwDvf")]').is_visible(timeout=5000):
                logging.info("Found single result page")
                results.append(extract_place_details(page))
                return results

            # Handle list of results
            listing_selector = '//a[contains(@href, "https://www.google.com/maps/place")]'
            try:
                page.wait_for_selector(listing_selector, timeout=10000)
            except:
                logging.warning("No listings found")
                return results

            # Scroll to load more results
            scroll_attempts = 0
            while page.locator(listing_selector).count() < limit and scroll_attempts < 5:
                scrollable = page.locator('div[role="feed"], div[aria-label*="Results for"]')
                if scrollable.count() > 0:
                    scrollable.first.evaluate("el => el.scrollTop += 5000")
                else:
                    page.mouse.wheel(0, 5000)
                time.sleep(2)
                scroll_attempts += 1

            listings = page.locator(listing_selector).all()[:limit]
            logging.info(f"Found {len(listings)} listings")
            
            for idx, listing in enumerate(listings):
                try:
                    listing.click()
                    page.wait_for_selector('//h1[contains(@class, "DUwDvf")]', timeout=5000)
                    time.sleep(0.5)
                    details = extract_place_details(page)
                    if details["name"]:
                        results.append(details)
                        logging.info(f"Extracted {idx+1}/{len(listings)}: {details['name']}")
                except Exception as e:
                    logging.warning(f"Failed to extract listing {idx+1}: {e}")
                    
        finally:
            browser.close()
    
    return results


def main():
    parser = argparse.ArgumentParser(description="Scrape business data from Google Maps")
    parser.add_argument("query", help="Search query for Google Maps")
    parser.add_argument("-l", "--limit", type=int, default=10, help="Maximum number of results (default: 10)")
    parser.add_argument("-o", "--output", help="Output JSON file path")
    parser.add_argument("-v", "--verbose", action="store_true", help="Enable verbose logging")
    parser.add_argument("--no-headless", action="store_false", dest="headless", help="Run browser in visible mode")
    
    args = parser.parse_args()
    setup_logging(args.verbose)
    
    try:
        results = scrape_google_maps(args.query, args.limit, args.headless)
        
        if args.output:
            with open(args.output, 'w') as f:
                json.dump(results, f, indent=2)
            logging.info(f"Saved {len(results)} results to {args.output}")
        else:
            print(json.dumps(results, indent=2))
        
        logging.info(f"Successfully scraped {len(results)} businesses")
        sys.exit(0)
    except Exception as e:
        logging.error(f"Error: {e}")
        sys.exit(1)


if __name__ == "__main__":
    main()

File: plugins/google-maps-scraper/scraper.js
// Google Maps Scraper using Lightpanda + puppeteer-core
// Extracts business data from Google Maps search results
// v0.3.0 - Updated consent flow + better extraction

const SKIP_WORDS = [
  'google', 'résultats', 'filtres', 'prix', 'disponible', 'inclus',
  'changer', 'invités', 'connecter', 'cookies', 'privacy', 'terms',
  'utiliser', 'sans frais', 'précédent', 'suivant', 'plan',
  'clavier', 'fléchées', 'déplacer', 'appuyer', 'numérique',
  'mercredi', 'jeudi', 'vendredi', 'samedi', 'dimanche', 'lundi', 'mardi',
  "aujourd'hui", 'demain', 'nuit', 'euro'
];

function isBusinessName(text) {
  if (!text || text.length < 3) return false;
  if (text.length > 120) return false;
  const lower = text.toLowerCase();
  for (const word of SKIP_WORDS) {
    if (lower.includes(word)) return false;
  }
  return true;
}

async function acceptConsent(page) {
  try {
    const hasConsent = await page.evaluate(() => {
      return document.body.innerText.includes("Avant d'accéder") ||
             document.body.innerText.includes('Before you continue');
    });
    if (!hasConsent) return false;

    console.log('Consent page detected, handling...');
    await new Promise(r => setTimeout(r, 2000));

    // Click "Plus d'options" / "More options"
    await page.evaluate(() => {
      for (const a of document.querySelectorAll('a')) {
        const t = a.textContent || '';
        if (t.includes("Plus d'options") || t.includes('More options')) {
          a.click(); return;
        }
      }
    });
    await new Promise(r => setTimeout(r, 2000));

    // Click "Tout refuser" / "Reject all"
    await page.evaluate(() => {
      for (const b of document.querySelectorAll('button')) {
        const t = (b.textContent || '').trim();
        if (t.includes('Tout refuser') || t.includes('Reject all')) {
          b.click(); return;
        }
      }
    });
    await new Promise(r => setTimeout(r, 4000));
    console.log('Consent handled (rejected personalization)');
    return true;
  } catch (e) {
    console.log('Consent handling error: ' + e.message);
    return false;
  }
}

async function scrapeGoogleMaps(query, limit) {
  console.log('Scraping Google Maps for: ' + query);

  const searchUrl = 'https://www.google.com/maps/search/' + query.replace(/ /g, '+');
  await page.goto(searchUrl, { waitUntil: 'domcontentloaded', timeout: 60000 });

  await acceptConsent(page);
  await new Promise(r => setTimeout(r, 3000));

  // Scroll to load more results
  for (let s = 0; s < 10; s++) {
    await page.mouse.wheel(0, 3000);
    await new Promise(r => setTimeout(r, 1500));
  }

  const rawListings = await page.evaluate(() => {
    const seen = new Set();
    const results = [];

    // Strategy 1: Place links with aria-label
    for (const el of document.querySelectorAll('a[href*="/maps/place/"]')) {
      const name = el.getAttribute('aria-label');
      if (name && !seen.has(name)) {
        seen.add(name);
        results.push({ method: 'place', name });
      }
    }

    // Strategy 2: aria-label divs with business-like names (all-caps or title case)
    for (const el of document.querySelectorAll('div[aria-label]')) {
      const label = el.getAttribute('aria-label') || '';
      if (!label || label.length < 3) continue;
      const firstLine = label.split('\n')[0].trim();
      if (firstLine && !seen.has(firstLine)) {
        if (/^[A-Z][A-Za-zÀ-ÿ\s'\-\d]+$/.test(firstLine) && firstLine.length > 3) {
          seen.add(firstLine);
          results.push({ method: 'aria', name: firstLine });
        }
      }
    }

    // Strategy 3: Sidebar listing cards
    for (const el of document.querySelectorAll('.Nv2PK, .THOPZb, .lI9IFe')) {
      const text = el.textContent || '';
      const lines = text.split('\n').map(l => l.trim()).filter(l => l);
      const name = lines[0];
      if (name && !seen.has(name) && name.length > 3) {
        seen.add(name);
        let rating = null, reviews = null;
        for (const line of lines) {
          const rm = line.match(/^([\d.]+)\s*[★☆]/);
          if (rm) {
            rating = parseFloat(rm[1]);
            const rcm = line.match(/\((\d+)\)/);
            if (rcm) reviews = parseInt(rcm[1], 10);
          }
        }
        results.push({ method: 'card', name, rating, reviews });
      }
    }

    return results;
  });

  console.log('Found ' + rawListings.length + ' raw listings');

  const seen = new Set();
  const results = [];
  for (const item of rawListings) {
    if (results.length >= limit) break;
    if (seen.has(item.name)) continue;
    seen.add(item.name);
    if (!isBusinessName(item.name)) continue;

    results.push({
      name: item.name,
      rating: item.rating || null,
      reviews_count: item.reviews || null,
      address: '',
      website: '',
      phone: '',
      type: '',
      query: query
    });
    console.log('Extracted ' + results.length + ': ' + item.name);
  }

  return results;
}

const query = context.args.query || "restaurants in Annecy France";
const limit = context.args.limit || 20;

const results = await scrapeGoogleMaps(query, limit);
console.log('Successfully scraped ' + results.length + ' businesses');

return results;
